systems/controllers/cp_traj_gen.cc

```diff
@@ -0,0 +1,397 @@
+
+namespace dairlib {
+namespace systems {
+
+CPTrajGenerator::CPTrajGenerator(RigidBodyTree<double> * tree,
+                                 double mid_foot_height,
+                                 double desired_final_foot_height,
+                                 double desired_final_vertical_foot_velocity,
+                                 double max_CoM_to_CP_dist,
+                                 double stance_duration_per_leg,
+                                 int left_stance_state,
+                                 int right_stance_state,
+                                 int left_foot_idx,
+                                 Eigen::Vector3d pt_on_left_foot,
+                                 int right_foot_idx,
+                                 Eigen::Vector3d pt_on_right_foot,
+                                 int pelvis_idx,
+                                 bool is_walking_position_control,
+                                 bool is_feet_collision_avoid,
+                                 bool is_using_predicted_com,
+                                 double cp_offset,
+                                 double center_line_offset) :
+    tree_(tree),
+    mid_foot_height_(mid_foot_height),
+    desired_final_foot_height_(desired_final_foot_height),
+    desired_final_vertical_foot_velocity_(desired_final_vertical_foot_velocity),
+    max_CoM_to_CP_dist_(max_CoM_to_CP_dist),
+    stance_duration_per_leg_(stance_duration_per_leg),
+    left_stance_(left_stance_state),
+    right_stance_(right_stance_state),
+    left_foot_idx_(left_foot_idx),
+    right_foot_idx_(right_foot_idx),
+    pt_on_left_foot_(pt_on_left_foot),
+    pt_on_right_foot_(pt_on_right_foot),
+    pelvis_idx_(pelvis_idx),
+    is_walking_position_control_(is_walking_position_control),
+    is_feet_collision_avoid_(is_feet_collision_avoid),
+    is_using_predicted_com_(is_using_predicted_com),
+    cp_offset_(cp_offset),
+    center_line_offset_(center_line_offset) {
+  this->set_name("cp_traj");
+
+  // Input/Output Setup
+  state_port_ = this->DeclareVectorInputPort(OutputVector<double>(
+                  tree->get_num_positions(),
+                  tree->get_num_velocities(),
+                  tree->get_num_actuators())).get_index();
+
+  fsm_port_ = this->DeclareVectorInputPort(
+                BasicVector<double>(1)).get_index();
+  if (is_using_predicted_com) {
+    com_port_ = this->DeclareAbstractInputPort("CoM_traj",
+        drake::Value<ExponentialPlusPiecewisePolynomial<double>> {}).get_index();
+  }
+  if (is_walking_position_control) {
+    fp_port_ = this->DeclareVectorInputPort(BasicVector<double>(2)).get_index();
+  }
+
+  this->DeclareAbstractOutputPort("cp_traj", &CPTrajGenerator::CalcTrajs);
+
+  // State variables inside this controller block
+  DeclarePerStepDiscreteUpdateEvent(&CPTrajGenerator::DiscreteVariableUpdate);
+  // The swing foot position at touchdown
+  prev_td_swing_foot_idx_ = this->DeclareDiscreteState(3);
```

Comment:
Isn't this just the stance foot position?

Reply:
It is both the stance foot position at the end of stance phase, and the swing foot position at the start of swing phase.

Reply:
Ah, that wasn't clear.